# Exercise 1

In [2]:
!pip install -q pyomo

     |████████████████████████████████| 8.9 MB 23.0 MB/s 
     |████████████████████████████████| 49 kB 5.7 MB/s 


In [6]:
from pyomo.environ import *

In [7]:
import numpy as np

In [8]:
model = ConcreteModel()

In [9]:
N = 12

In [12]:
obj_coeff = np.array([ -8.1, 10.15, 30.5, 50.05, 0.05, 80.5, -0.25, -31.02, 50.65, 0.725, -0.8, 100.6 ])

In [13]:
constr1_coeff = np.array([1,1,1,1,1,1,1,1,1,1,1,1])
constr2_coeff = np.array([0, 1.6, -2.02, 0, 1.01, -2.005, 0, 1.39, 1, -1.214, 0, -5.32])

In [14]:
constr1_rhs = 161.2
constr2_rhs = 52.6

In [15]:
lower_bound = np.array([-np.inf, 1, 0, 0, 7, 0, -np.inf, -np.inf, 1, 1, -5, 2])
upper_bound = np.array([4,3,np.inf,2,10,np.inf,13,20,np.inf,21,5,60])

In [16]:
col_indices = np.arange(N)

In [17]:
model.x = Var(col_indices)

In [18]:
model.constraints = ConstraintList()

In [19]:
model.constraints.add(sum(constr1_coeff[i]*model.x[i] for i in col_indices) <= constr1_rhs)

In [20]:
model.constraints.add(sum(constr2_coeff[i]*model.x[i] for i in col_indices) <= constr2_rhs)

In [22]:
for i in col_indices:
  model.x[i].setlb(lower_bound[i])
  model.x[i].setub(upper_bound[i])

In [23]:
model.obj = Objective(expr = sum(obj_coeff[i]*model.x[i] for i in col_indices), sense = minimize)

In [24]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    2 : {1, 2}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   12 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11}

1 Var Declarations
    x : Size=12, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :  -inf :  None :   4.0 : False :  True :  Reals
          1 :   1.0 :  None :   3.0 : False :  True :  Reals
          2 :   0.0 :  None :   inf : False :  True :  Reals
          3 :   0.0 :  None :   2.0 : False :  True :  Reals
          4 :   7.0 :  None :  10.0 : False :  True :  Reals
          5 :   0.0 :  None :   inf : False :  True :  Reals
          6 :  -inf :  None :  13.0 : False :  True :  Reals
          7 :  -inf :  None :  20.0 : False :  True :  Reals
          8 :   1.0 :  None :   inf : False :  True :

In [25]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160837 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

In [26]:
SolverFactory('glpk', executable = '/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: -396.975
  Upper bound: -396.975
  Number of objectives: 1
  Number of constraints: 3
  Number of variables: 13
  Number of nonzeros: 21
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.012834310531616211
# ----------------------------------------------------------
#   Solution Information
# ----------------------------

In [27]:
print('Objective = ', model.obj())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()

Objective =  -396.97499999999997

Decision Variables
x[ 0 ] =  4.0
x[ 1 ] =  1.0
x[ 2 ] =  0.0
x[ 3 ] =  0.0
x[ 4 ] =  7.0
x[ 5 ] =  0.0
x[ 6 ] =  13.0
x[ 7 ] =  20.0
x[ 8 ] =  1.0
x[ 9 ] =  1.0
x[ 10 ] =  5.0
x[ 11 ] =  2.0

Constraints
constraints : Size=2
    Key : Lower : Body   : Upper
      1 :  None :   54.0 : 161.2
      2 :  None : 25.616 :  52.6


## Ans 2 :

The optimal solution is -396.975 at which the values of decision variables are as follows:

$x_1 = 4.0\\x_2 = 1.0\\x_3 = 0.0\\x_4 = 0.0\\x_5 = 7.0\\x_6 = 0.0\\x_7 = 13.0\\x_8 = 20.0\\x_9 = 1.0\\x_{10} = 1.0\\x_{11} = 5.0\\x_{12} = 2.0$

The value of the objective function is -396.975.